account-compression/sdk/tests/accountCompression.test.ts

```diff
@@ -0,0 +1,684 @@
+        "CMT updated its active index after attacker's transaction, when it shouldn't have done anything"
+      );
+    });
+    it("Random attacker fails to fake the existence of a leaf by autocompleting proof", async () => {
+      const maliciousLeafHash = crypto.randomBytes(32);
+      const maliciousLeafHash1 = crypto.randomBytes(32);
+      const nodeProof: Buffer[] = [];
+      for (let i = 0; i < DEPTH; i++) {
+        nodeProof.push(Buffer.alloc(32));
+      }
+
+      // Root - make this nonsense so it won't match what's in CL, and force proof autocompletion
+      const replaceIx = createReplaceIx(
+        payer,
+        splCMTKeypair.publicKey,
+        Buffer.alloc(32),
+        maliciousLeafHash,
+        maliciousLeafHash1,
+        0,
+        nodeProof
+      );
+
+      try {
+        await execute(provider, [replaceIx], [payer]);
+        assert(
+          false,
+          "Attacker was able to succesfully write fake existence of a leaf"
+        );
+      } catch (e) { }
+
+      const splCMT = await getConcurrentMerkleTree(provider.connection, splCMTKeypair.publicKey);
+
+      assert(
+        getCMTActiveIndex(splCMT) === 0,
+        "CMT updated its active index after attacker's transaction, when it shouldn't have done anything"
+      );
+    });
+  });
+  describe(`Canopy test`, () => {
+    const DEPTH = 5;
+    it("Testing canopy for appends and replaces on a full on chain tree", async () => {
+      [splCMTKeypair, offChainTree] = await createTreeOnChain(
+        payer,
+        0,
+        DEPTH,
+        8,
+        DEPTH // Store full tree on chain
+      );
+
+      let leaves: Array<number>[] = [];
+      let i = 0;
+      let stepSize = 4;
+      while (i < 2 ** DEPTH) {
+        let ixs: TransactionInstruction[] = [];
+        for (let j = 0; j < stepSize; ++j) {
+          const newLeaf = Array.from(Buffer.alloc(32, i + 1));
+          leaves.push(newLeaf);
+          const appendIx = createAppendIx(
+            newLeaf,
+            payer,
+            splCMTKeypair.publicKey
+          );
+          ixs.push(appendIx);
+        }
+        await execute(provider, ixs, [payer]);
+        i += stepSize;
+        console.log("Appended", i, "leaves");
+      }
+
+      // Compare on-chain & off-chain roots
+      let ixs: TransactionInstruction[] = [];
+      const splCMT = await getConcurrentMerkleTree(connection, splCMTKeypair.publicKey);
+      const root = getCMTCurrentRoot(splCMT);
+
+      let leafList = Array.from(leaves.entries());
+      leafList.sort(() => Math.random() - 0.5);
+      let replaces = 0;
+      let newLeaves = {};
+      for (const [i, leaf] of leafList) {
+        const newLeaf = crypto.randomBytes(32);
+        newLeaves[i] = newLeaf;
+        const replaceIx = createReplaceIx(
+          payer,
+          splCMTKeypair.publicKey,
+          root,
+          Buffer.from(leaf),
+          newLeaf,
+          i,
+          [] // No proof necessary
```

Comment:
Since I'm still a noob, I have to ask: why is no proof necessary? Is it because everything's at the lowest level in the tree, and not yet pushed up higher? I'm not even sure if I'm using the right terminology

Reply:
Not a noob question. I added documentation to explain this test.

This test puts the entire merkle tree on-chain by storing it entirely in the `canopy` section. Since the `canopy` stores the full state of the tree, we are actually able to infer the proof to any leaf node, so it does not need to be passed in the instruction.